### PR TITLE
Added lesson 8

### PR DESCRIPTION
**Pull request by luisesch**

### CODE REVIEW

**Comment by pektinasen** [comment]
I wrote  a whole essay about why I think we should teach for loops first and then transition to while loops, but then I clicked somewhere and all my text was gone 😭. so here's the gist:

* first we show all these examples with while loops, and then we say: Actually, we don't use while loops in these cases
* there is hardly any use case in modern js where a while loop expresses more clearly what we want to do than a for loop
* there are many more for loops that aren't even introduced here
* i don't think we have a single while loop in our react-native project at work
* examples for while loops:
** IO: reading/writing bytes from/to the wire
** game loops: while (keepPlaying) { newGame(); }
** animations

`var test = true`

1. `for (;test;) { test = false}`
2. `while (test) { test = false}`

1 and 2 are equivalent.

git related: for future PRs, could you try to not commit whitespace changes. I find it hard to see if something meaningful changed in index.html.

The last weeks made me realize how opiniated I am about coding. Just to clarify, I'm not saying we should do what I'm saying. It's always meant as a suggestion or a start of a discussion. Please let me know if anything comes off rude. It's not meant that way (Intention does not change the effect, though).

**Comment by luisesch** [comment]
First of all, thank you for the detailed feedback! I think your comments are always very helpful and valuable and don't come off as rude at all.
I definitely agree with what you're writing, however, here is my essay on why I think from a learner's perspective it still does make sense to explain while loops before turning to for loops :)
I think that the syntax of a while loop is a lot more intuitive to someone first getting in touch with loops than the rather complex (and at first possibly confusing) way in which for loops are built. So I think for someone who has never heard of loops before, while loops might make it easier to get a general understanding of the whole loop concept, how they work and what they do.
Looking at the three different parts/steps of a while loop and then transfering them to a for loop I thought was a very smart way of approaching the topic.
Regarding other types of for loops, I think it might make more sense to teach them once we have covered some more data types (like objects and arrays), also in order not to overwhelm the students.

But I'd definitely also be interested in hearing how the other teachers feel about this question.

And finally, I would love to not commit whitespace changes! Do you have any advice on how? :)

**Comment by pektinasen** [comment]
Definitely makes sense what you are saying, that they are easier to understand because the mechanism is simpler.
I'm just concerned that they use the while loop for everything because it's the tool that was first introduced to them and the most time was spent on.

Re git and formatting: I think the first issue is that you are using different formatting settings for html than the original committer. And you have somehow autoformatted the whole file.

And for git, I usually use `git add -p` when staging files. Which basically is a first code review by myself before committing. Jetbrains Ides can also stage only selected hunks.

Maybe you have a git setting that doesn't show you whitespace differences in a diff.

**Comment by haraldF** [comment]
My 2 cents :) My code is full of `while`, mostly because I'm a fan of capturing regexes :) I checked some of our internal TypeScript code and the ratio of `while` / `for` is 1/7, so yes, while does have a reason of existence...

I agree with Luise that while is more intuitive and since we can use the generic `while` to explain `for`, I vote for introducing it first.

File: lesson8.md [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+  i++;
+}
+```
+
+Squaring the first 10 numbers
+
+```js
+let i = 1;
+while (i < 11) {
+  console.log(i * i);
```
Comment (haraldF) [inline lesson8.md]:
```suggestion
  console.log(i * i); // or i ** 2
```

**Pull request merged**